app/models/manager_refresh/inventory_collection.rb

```diff
@@ -146,6 +149,11 @@ def object_index_with_keys(keys, object)
       keys.map { |attribute| object.public_send(attribute).to_s }.join(stringify_joiner)
     end
 
+    def find_or_build(manager_uuid)
+      # FIXME: splat manager_ref
+      data_index[manager_uuid] || build({ manager_ref.first => manager_uuid })
```

Comment:
the exception could nicely say, allowed find_by values are {:ems_ref => xy}, {:hardware => xy, :device_name => xx}